core/Enclave/key_operation.cpp

```diff
@@ -0,0 +1,2038 @@
+    case EH_SHA_2_224:
+        return EVP_sha224();
+    case EH_SHA_2_256:
+        return EVP_sha256();
+    case EH_SHA_2_384:
+        return EVP_sha384();
+    case EH_SHA_2_512:
+        return EVP_sha512();
+    case EH_SM3:
+        return EVP_sm3();
+    default:
+        return NULL;
+    }
+}
+
+/**
+ * @brief Check parameters and encrypted data
+ * @param aad Additional data
+ * @param cmk Key information
+ * @param plaintext Data to be encrypted
+ * @param cipherblob The information of ciphertext
+ */
+sgx_status_t ehsm_aes_gcm_encrypt(const ehsm_data_t *aad,
+                                  const ehsm_keyblob_t *cmk,
+                                  const ehsm_data_t *plaintext,
+                                  ehsm_data_t *cipherblob)
+{
+    sgx_status_t ret = SGX_ERROR_UNEXPECTED;
+
+    int temp_len = 0;
+    EVP_CIPHER_CTX *pctx = NULL;
+
+    /* this api only support for symmetric keys */
+    if (cmk->metadata.keyspec != EH_AES_GCM_128 &&
+        cmk->metadata.keyspec != EH_AES_GCM_192 &&
+        cmk->metadata.keyspec != EH_AES_GCM_256)
+    {
+        return SGX_ERROR_INVALID_PARAMETER;
+    }
+
+    /* calculate the ciphertext length */
+    if (cipherblob->datalen == 0)
+    {
+        cipherblob->datalen = plaintext->datalen + EH_AES_GCM_IV_SIZE + EH_AES_GCM_MAC_SIZE;
+        return SGX_SUCCESS;
+    }
+
+    uint32_t keysize = 0;
+    if (!ehsm_get_symmetric_key_size(cmk->metadata.keyspec, keysize))
+    {
+        return SGX_ERROR_UNEXPECTED;
+    }
+
+    uint32_t enc_key_size = ehsm_get_gcm_ciphertext_size((sgx_aes_gcm_data_ex_t *)cmk->keyblob);
+    if (enc_key_size == UINT32_MAX || enc_key_size != keysize)
+    {
+        log_d("enc_key_size:%d is not expected: %lu.\n", enc_key_size, keysize);
+        return SGX_ERROR_INVALID_PARAMETER;
+    }
+
+    if (plaintext->datalen > EH_ENCRYPT_MAX_SIZE)
+        return SGX_ERROR_INVALID_PARAMETER;
+
+    if (cipherblob->datalen < plaintext->datalen + SGX_AESGCM_IV_SIZE + SGX_AESGCM_MAC_SIZE)
+        return SGX_ERROR_INVALID_PARAMETER;
+
+    uint8_t *iv = (uint8_t *)(cipherblob->data + plaintext->datalen);
+    uint8_t *mac = (uint8_t *)(cipherblob->data + plaintext->datalen + SGX_AESGCM_IV_SIZE);
+    uint8_t *enc_key = (uint8_t *)malloc(keysize);
+    if (enc_key == NULL)
+    {
+        return SGX_ERROR_OUT_OF_MEMORY;
+    }
+
+    const EVP_CIPHER *block_mode = get_symmetric_block_mode(cmk->metadata.keyspec);
+    if (block_mode == NULL)
+    {
+        ret = SGX_ERROR_UNEXPECTED;
+        goto out;
+    }
+
+    ret = sgx_read_rand(iv, SGX_AESGCM_IV_SIZE);
+    if (ret != SGX_SUCCESS)
+    {
+        log_d("error generating IV\n");
+        goto out;
+    }
+
+    ret = ehsm_parse_keyblob(enc_key, enc_key_size,
+                             (sgx_aes_gcm_data_ex_t *)cmk->keyblob);
+    if (ret != SGX_SUCCESS)
+    {
+        log_d("failed to decrypt key\n");
+        goto out;
+    }
+
+    // Create and init ctx
+    if (!(pctx = EVP_CIPHER_CTX_new()))
```

Comment:
from line 203 to 253, put into a new function, which can also be used in self test.